Introduce a small program to simplify a Node's log

abailly-iohk opened the pull request:
This PR adds an executable `log-filter` that can be used to, well, filter logs emitted by Hydra node to only get the "interesting" parts, namely the events and effects. This small tool should make our lives easier when troubleshooting issues with nodes, and we could extend it to do more interesting things later, like analysis or monitoring, removing the Prometheus server from the HydraNode proper.

abailly committed:
Introduce an executable to simplify logs

abailly committed:
Simplify more events and effects

abailly committed:
Allow filtering standard input

abailly committed:
Properly filter transactions and add timestamp

abailly committed:
Add options parsing and some help

KtorZ commented on local-cluster/src/Hydra/LogFilter.hs:
```diff
@@ -0,0 +1,82 @@
+filterLog entry = do
+  guard (entry ^? key "message" . key "tag" == Just "Node")
+  node <- entry ^? key "message" . key "node"
+  result <- case node ^? key "tag" of
```
So you mean I can throw away my jq queries :smile: ?

```
jq '.message.node?|select(.tag? == "ProcessedEffect")|.effect?.serverOutput?|select(.tag == "SnapshotConfirmed")|.snapshot.confirmedTransactions|map(.id)
```

KtorZ commented on local-cluster/src/Hydra/LogFilter.hs:
```diff
@@ -0,0 +1,82 @@
+        Just "ReqTx" -> replaceTransactionWithTxId node networkEffect
+        Just "ReqSn" -> replaceTransactionsWithTxIds node networkEffect
+        _ -> pure node
+    Just "ProcessedEffect" -> do
```
So far, I've rarely been interested in both `ProcessingXXX` and `ProcessedXXX`. I usually find them redundant in terms of information. Perhaps we could actually assign effects a short identifier and simply refer to them in the `ProcessedXXX` kind of events to avoid duplicating information and bloating logs even more.

Just a thought...

KtorZ commented on local-cluster/src/Hydra/LogFilter.hs:
```diff
@@ -0,0 +1,82 @@
+  clientEvent = key "event" . key "clientInput"
+
+  clientEffect :: Traversal' Value Value
+  clientEffect = key "effect" . key "serverOutput"
```
I wonder about the maintainability of all this though, when we change fields and whatnot; Since it's all JSON-based, we don't get any compiler warning / errors in case we alter the JSON logs and render these traversals invalid.

KtorZ commented on local-cluster/test/Test/LogFilterSpec.hs:
```diff
@@ -0,0 +1,32 @@
+
+  it "replaces transactions by their ids in ReqSn" $ do
+    (decode entry >>= filterLog >>= (^? key "message" . key "event" . key "message" . key "transactions"))
+      `shouldBe` Just (Array [String "6cba5394ec8a1a1161758a33089661383143283d0121e4a293ed51a0272cfbc4"])
```
A property test idea: 

- Generate some arbitrary logs
- Measure their size unfiltered
- Measure their size filtered
- Use `cover` and `checkCoverage` to ensure that the "compression" is around some minimum acceptable value. 

This should perhaps help to catch field changes fairly well and makes maintenance slightly easier?

abailly committed:
Add coverage check for compression ratio of filtered logs

The idea is to check the log filter actually does something which is a
proxy for checking we correctly parse and interpret the log
messages. Ideally we would rather like to check the filter based on
the Log's schema to catch regressions, but that's somewhat more
involved.

abailly committed:
Simplify Monitoring test to make it less flacky

ghost commented:
\o/

ghost commented:
As discussed last week, I will merge this as is as it's pretty much independent and provides some useful tool for analysing logs.
